### PR TITLE
feat(cosmwasm): introduce `union-ibc`

### PR DESCRIPTION
**Pull request by hussein-aitlahcen**

### CODE REVIEW

File: cosmwasm/union-ibc/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,1486 @@
...
+    let connection_id = NEXT_CONNECTION_ID
+        .may_load(deps.storage)?
+        .unwrap_or_default();
+    NEXT_CONNECTION_ID.save(
+        deps.storage,
+        &connection_id
+            .checked_add(1)
+            .ok_or(ContractError::ArithmeticOverflow)?,
+    )?;
+    Ok(connection_id)
```
Comment (benluelo) [inline cosmwasm/union-ibc/src/contract.rs:1346]: is there no "increment" functionality available?

File: lib/ibc-solidity/src/cosmwasm/types.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
+// Duplicated until we resolve whether we reuse the same types, or not.
+
+pub mod ibc {
+    alloy::sol! {
+        // #![sol(cfg_attr(feature = "rpc", rpc))]
+        #![sol(rpc, all_derives)]
```
Comment (benluelo) [inline lib/ibc-solidity/src/cosmwasm/types.rs:6]: you can't use the rpc attr in cosmwasm contracts

File: lib/unionlabs/src/ics24.rs [hunk 3]
```diff
@@ -77,48 +77,48 @@ pub mod ethabi {
...
     #[must_use]
-    pub fn next_seq_send_key(channel_id: u32) -> H256 {
+    pub fn next_seq_ack_key(channel_id: u32) -> H256 {
         Keccak256::new()
-            .chain_update(NEXT_SEQ_SEND.to_be_bytes())
+            .chain_update(NEXT_SEQ_ACK.to_be_bytes())
             .chain_update(U256::from(channel_id).to_be_bytes())
             .finalize()
             .into()
     }
```
Comment (benluelo) [inline lib/unionlabs/src/ics24.rs:104]: these can be removed now yeah? since we no longer have sequence

**Review by benluelo — COMMENTED** [review]
lfg

general comments:

- string event keys should be pulled out into constants
- backlinks to the spec once that's finished

File: cosmwasm/union-ibc/src/contract.rs [hunk 4]
```diff
@@ -0,0 +1,1486 @@
...
+    Ok(())
+}
+
+fn commit_packet(packet: &Packet) -> H256 {
+    commit(packet.abi_encode())
+}
+
+fn commit_packets(packets: &[Packet]) -> H256 {
+    commit(packets.abi_encode())
+}
```
[the remark this answers is not in the document]
Comment (hussein-aitlahcen) [reply cosmwasm/union-ibc/src/contract.rs:533]: I think it's correct, we encode the array directly

File: lib/unionlabs/src/ics24.rs [hunk 5]
```diff
@@ -148,12 +118,6 @@ pub enum Path {
     #[display(fmt = "{_0}")]
     Receipt(ReceiptPath),
     #[display(fmt = "{_0}")]
-    NextSequenceSend(NextSequenceSendPath),
-    #[display(fmt = "{_0}")]
-    NextSequenceRecv(NextSequenceRecvPath),
-    #[display(fmt = "{_0}")]
-    NextSequenceAck(NextSequenceAckPath),
-    #[display(fmt = "{_0}")]
```
Comment (benluelo) [inline lib/unionlabs/src/ics24.rs:156]: this needs to stay because its the ics24 path, not our ibc spec path

File: cosmwasm/union-ibc/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+workspace = true
+
+[lib]
+crate-type = ["cdylib", "rlib"]
+
+[features]
+library = []
+
+[dependencies]
+alloy           = { workspace = true, features = ["sol-types", "contract"] }
```
Comment (benluelo) [inline cosmwasm/union-ibc/Cargo.toml:19]:
```suggestion
alloy           = { workspace = true, features = ["sol-types"] }
```

Comment (benluelo) [reply cosmwasm/union-ibc/Cargo.toml:19]: will fix compilation error about getrandom

**Review by benluelo — DISMISSED** [review]
🚀

**Review by benluelo — APPROVED** [review]

**Review by benluelo — APPROVED** [review]

**Pull request merged**